// javascript-time-ago/commonjs/gradation/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "canonical", {
  enumerable: true,
  get: function get() {
    return _canonical.default;
  }
});
Object.defineProperty(exports, "convenient", {
  enumerable: true,
  get: function get() {
    return _convenient.default;
  }
});
Object.defineProperty(exports, "minute", {
  enumerable: true,
  get: function get() {
    return _helpers.minute;
  }
});
Object.defineProperty(exports, "hour", {
  enumerable: true,
  get: function get() {
    return _helpers.hour;
  }
});
Object.defineProperty(exports, "day", {
  enumerable: true,
  get: function get() {
    return _helpers.day;
  }
});
Object.defineProperty(exports, "month", {
  enumerable: true,
  get: function get() {
    return _helpers.month;
  }
});
Object.defineProperty(exports, "year", {
  enumerable: true,
  get: function get() {
    return _helpers.year;
  }
});
Object.defineProperty(exports, "getStep", {
  enumerable: true,
  get: function get() {
    return _helpers.getStep;
  }
});
Object.defineProperty(exports, "getDate", {
  enumerable: true,
  get: function get() {
    return _helpers.getDate;
  }
});

var _canonical = _interopRequireDefault(require("./canonical"));

var _convenient = _interopRequireDefault(require("./convenient"));

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
